### PR TITLE
EdgeHub: Close connection when all links are closed

### PR DESCRIPTION
**Pull request by varunpuranik**
For AMQP, a "connection" on the EdgeHub side consists of multiple links on the AMQP Protocol head side. So a connection is considered opened when any one link is opened for a client. Similarly, we need to consider a connection closed when all links for a client are closed.

### CODE REVIEW

**Comment by myagley** [comment]
How did we find this and what are the symptoms?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Amqp/ConnectionHandler.cs [hunk 1]
```diff
@@ -154,6 +154,31 @@ public async Task RegisterLinkHandler(ILinkHandler linkHandler)
...
+                    {
+                        await this.CloseConnection();
+                    }
+                }
+            }
+        }
+
+        async Task CloseConnection()
+        {
+            using (await this.initializationLock.LockAsync())
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Amqp/ConnectionHandler.cs]: There's two locks in play here. Is there any chance for a deadlock?

**Comment by varunpuranik** [comment]
This is not a bug as such, and there is no functionality being affected. This is just to get the logging right.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Amqp/ConnectionHandler.cs (hunk 1 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Amqp/ConnectionHandler.cs]: No, because the 2 locks are not used together in the other case.

**Pull request merged**